{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bedtId8GreAh"
      },
      "source": [
        "# Aula Prática: Alinhando uma Amostra de E. coli a um Genoma de Referência\n",
        "\n",
        "## Ferramentas que usaremos:\n",
        "\n",
        "* Google Colab: Para executar a análise na nuvem.\n",
        "\n",
        "* Biopython: Para manipular dados de sequência.\n",
        "\n",
        "* Bowtie2: Para alinhamento.\n",
        "\n",
        "* SAMtools: Para processar arquivos de alinhamento.\n",
        "\n",
        "* Pysam: Para visualizar resultados de alinhamento."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OppZ7y5-rKMr",
        "outputId": "419d12e7-0eb9-448e-c849-46cf655d60cb"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Hit:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease\n",
            "Hit:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease\n",
            "Hit:3 http://security.ubuntu.com/ubuntu jammy-security InRelease\n",
            "Hit:4 https://r2u.stat.illinois.edu/ubuntu jammy InRelease\n",
            "Hit:5 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "Hit:6 http://archive.ubuntu.com/ubuntu jammy-updates InRelease\n",
            "Hit:7 http://archive.ubuntu.com/ubuntu jammy-backports InRelease\n",
            "Hit:8 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n",
            "Hit:9 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n",
            "Hit:10 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "32 packages can be upgraded. Run 'apt list --upgradable' to see them.\n",
            "\u001b[1;33mW: \u001b[0mSkipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\u001b[0m\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "bowtie2 is already the newest version (2.4.4-1).\n",
            "samtools is already the newest version (1.13-4).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 32 not upgraded.\n",
            "Requirement already satisfied: biopython in /usr/local/lib/python3.11/dist-packages (1.85)\n",
            "Requirement already satisfied: pysam in /usr/local/lib/python3.11/dist-packages (0.23.0)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from biopython) (2.0.2)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "Suggested packages:\n",
            "  python3-numpy python3-matplotlib texlive-latex-recommended\n",
            "The following NEW packages will be installed:\n",
            "  bcftools\n",
            "0 upgraded, 1 newly installed, 0 to remove and 32 not upgraded.\n",
            "Need to get 697 kB of archives.\n",
            "After this operation, 2,360 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 bcftools amd64 1.13-1 [697 kB]\n",
            "Fetched 697 kB in 1s (701 kB/s)\n",
            "Selecting previously unselected package bcftools.\n",
            "(Reading database ... 125519 files and directories currently installed.)\n",
            "Preparing to unpack .../bcftools_1.13-1_amd64.deb ...\n",
            "Unpacking bcftools (1.13-1) ...\n",
            "Setting up bcftools (1.13-1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n"
          ]
        }
      ],
      "source": [
        "# Instalação das ferramentas necessárias\n",
        "!apt update\n",
        "!apt install -y bowtie2 samtools\n",
        "!pip install biopython pysam\n",
        "!apt install -y bcftools"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Y01nVGYpsdnO",
        "outputId": "1bbba1da-b6ef-4bee-e4aa-b7408aa4256a"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "--2025-03-17 20:02:35--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz\n",
            "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.12, 130.14.250.13, 130.14.250.31, ...\n",
            "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.12|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1379902 (1.3M) [application/x-gzip]\n",
            "Saving to: ‘GCF_000005845.2_ASM584v2_genomic.fna.gz’\n",
            "\n",
            "GCF_000005845.2_ASM 100%[===================>]   1.32M  4.77MB/s    in 0.3s    \n",
            "\n",
            "2025-03-17 20:02:38 (4.77 MB/s) - ‘GCF_000005845.2_ASM584v2_genomic.fna.gz’ saved [1379902/1379902]\n",
            "\n",
            "Settings:\n",
            "  Output files: \"ecoli_ref.*.bt2\"\n",
            "  Line rate: 6 (line is 64 bytes)\n",
            "  Lines per side: 1 (side is 64 bytes)\n",
            "  Offset rate: 4 (one in 16)\n",
            "  FTable chars: 10\n",
            "  Strings: unpacked\n",
            "  Max bucket size: default\n",
            "  Max bucket size, sqrt multiplier: default\n",
            "  Max bucket size, len divisor: 4\n",
            "  Difference-cover sample period: 1024\n",
            "  Endianness: little\n",
            "  Actual local endianness: little\n",
            "  Sanity checking: disabled\n",
            "  Assertions: disabled\n",
            "  Random seed: 0\n",
            "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
            "Input files DNA, FASTA:\n",
            "  ecoli_ref.fna\n",
            "Building a SMALL index\n",
            "Reading reference sizes\n",
            "  Time reading reference sizes: 00:00:01\n",
            "Calculating joined length\n",
            "Writing header\n",
            "Reserving space for joined string\n",
            "Joining reference sequences\n",
            "  Time to join reference sequences: 00:00:00\n",
            "bmax according to bmaxDivN setting: 1160413\n",
            "Using parameters --bmax 870310 --dcv 1024\n",
            "  Doing ahead-of-time memory usage test\n",
            "  Passed!  Constructing with these parameters: --bmax 870310 --dcv 1024\n",
            "Constructing suffix-array element generator\n",
            "Building DifferenceCoverSample\n",
            "  Building sPrime\n",
            "  Building sPrimeOrder\n",
            "  V-Sorting samples\n",
            "  V-Sorting samples time: 00:00:00\n",
            "  Allocating rank array\n",
            "  Ranking v-sort output\n",
            "  Ranking v-sort output time: 00:00:00\n",
            "  Invoking Larsson-Sadakane on ranks\n",
            "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
            "  Sanity-checking and returning\n",
            "Building samples\n",
            "Reserving space for 12 sample suffixes\n",
            "Generating random suffixes\n",
            "QSorting 12 sample offsets, eliminating duplicates\n",
            "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
            "Multikey QSorting 12 samples\n",
            "  (Using difference cover)\n",
            "  Multikey QSorting samples time: 00:00:00\n",
            "Calculating bucket sizes\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 1, merged 6; iterating...\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Avg bucket size: 663092 (target: 870309)\n",
            "Converting suffix-array elements to index image\n",
            "Allocating ftab, absorbFtab\n",
            "Entering Ebwt loop\n",
            "Getting block 1 of 7\n",
            "  Reserving size (870310) for bucket 1\n",
            "  Calculating Z arrays for bucket 1\n",
            "  Entering block accumulator loop for bucket 1:\n",
            "  bucket 1: 10%\n",
            "  bucket 1: 20%\n",
            "  bucket 1: 30%\n",
            "  bucket 1: 40%\n",
            "  bucket 1: 50%\n",
            "  bucket 1: 60%\n",
            "  bucket 1: 70%\n",
            "  bucket 1: 80%\n",
            "  bucket 1: 90%\n",
            "  bucket 1: 100%\n",
            "  Sorting block of length 415161 for bucket 1\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:01\n",
            "Returning block of 415162 for bucket 1\n",
            "Getting block 2 of 7\n",
            "  Reserving size (870310) for bucket 2\n",
            "  Calculating Z arrays for bucket 2\n",
            "  Entering block accumulator loop for bucket 2:\n",
            "  bucket 2: 10%\n",
            "  bucket 2: 20%\n",
            "  bucket 2: 30%\n",
            "  bucket 2: 40%\n",
            "  bucket 2: 50%\n",
            "  bucket 2: 60%\n",
            "  bucket 2: 70%\n",
            "  bucket 2: 80%\n",
            "  bucket 2: 90%\n",
            "  bucket 2: 100%\n",
            "  Sorting block of length 704345 for bucket 2\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:00\n",
            "Returning block of 704346 for bucket 2\n",
            "Getting block 3 of 7\n",
            "  Reserving size (870310) for bucket 3\n",
            "  Calculating Z arrays for bucket 3\n",
            "  Entering block accumulator loop for bucket 3:\n",
            "  bucket 3: 10%\n",
            "  bucket 3: 20%\n",
            "  bucket 3: 30%\n",
            "  bucket 3: 40%\n",
            "  bucket 3: 50%\n",
            "  bucket 3: 60%\n",
            "  bucket 3: 70%\n",
            "  bucket 3: 80%\n",
            "  bucket 3: 90%\n",
            "  bucket 3: 100%\n",
            "  Sorting block of length 715232 for bucket 3\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:00\n",
            "Returning block of 715233 for bucket 3\n",
            "Getting block 4 of 7\n",
            "  Reserving size (870310) for bucket 4\n",
            "  Calculating Z arrays for bucket 4\n",
            "  Entering block accumulator loop for bucket 4:\n",
            "  bucket 4: 10%\n",
            "  bucket 4: 20%\n",
            "  bucket 4: 30%\n",
            "  bucket 4: 40%\n",
            "  bucket 4: 50%\n",
            "  bucket 4: 60%\n",
            "  bucket 4: 70%\n",
            "  bucket 4: 80%\n",
            "  bucket 4: 90%\n",
            "  bucket 4: 100%\n",
            "  Sorting block of length 638686 for bucket 4\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:00\n",
            "Returning block of 638687 for bucket 4\n",
            "Getting block 5 of 7\n",
            "  Reserving size (870310) for bucket 5\n",
            "  Calculating Z arrays for bucket 5\n",
            "  Entering block accumulator loop for bucket 5:\n",
            "  bucket 5: 10%\n",
            "  bucket 5: 20%\n",
            "  bucket 5: 30%\n",
            "  bucket 5: 40%\n",
            "  bucket 5: 50%\n",
            "  bucket 5: 60%\n",
            "  bucket 5: 70%\n",
            "  bucket 5: 80%\n",
            "  bucket 5: 90%\n",
            "  bucket 5: 100%\n",
            "  Sorting block of length 779115 for bucket 5\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:01\n",
            "Returning block of 779116 for bucket 5\n",
            "Getting block 6 of 7\n",
            "  Reserving size (870310) for bucket 6\n",
            "  Calculating Z arrays for bucket 6\n",
            "  Entering block accumulator loop for bucket 6:\n",
            "  bucket 6: 10%\n",
            "  bucket 6: 20%\n",
            "  bucket 6: 30%\n",
            "  bucket 6: 40%\n",
            "  bucket 6: 50%\n",
            "  bucket 6: 60%\n",
            "  bucket 6: 70%\n",
            "  bucket 6: 80%\n",
            "  bucket 6: 90%\n",
            "  bucket 6: 100%\n",
            "  Sorting block of length 548871 for bucket 6\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:00\n",
            "Returning block of 548872 for bucket 6\n",
            "Getting block 7 of 7\n",
            "  Reserving size (870310) for bucket 7\n",
            "  Calculating Z arrays for bucket 7\n",
            "  Entering block accumulator loop for bucket 7:\n",
            "  bucket 7: 10%\n",
            "  bucket 7: 20%\n",
            "  bucket 7: 30%\n",
            "  bucket 7: 40%\n",
            "  bucket 7: 50%\n",
            "  bucket 7: 60%\n",
            "  bucket 7: 70%\n",
            "  bucket 7: 80%\n",
            "  bucket 7: 90%\n",
            "  bucket 7: 100%\n",
            "  Sorting block of length 840236 for bucket 7\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:00\n",
            "Returning block of 840237 for bucket 7\n",
            "Exited Ebwt loop\n",
            "fchr[A]: 0\n",
            "fchr[C]: 1142742\n",
            "fchr[G]: 2322833\n",
            "fchr[T]: 3500270\n",
            "fchr[$]: 4641652\n",
            "Exiting Ebwt::buildToDisk()\n",
            "Returning from initFromVector\n",
            "Wrote 5741800 bytes to primary EBWT file: ecoli_ref.1.bt2\n",
            "Wrote 1160420 bytes to secondary EBWT file: ecoli_ref.2.bt2\n",
            "Re-opening _in1 and _in2 as input streams\n",
            "Returning from Ebwt constructor\n",
            "Headers:\n",
            "    len: 4641652\n",
            "    bwtLen: 4641653\n",
            "    sz: 1160413\n",
            "    bwtSz: 1160414\n",
            "    lineRate: 6\n",
            "    offRate: 4\n",
            "    offMask: 0xfffffff0\n",
            "    ftabChars: 10\n",
            "    eftabLen: 20\n",
            "    eftabSz: 80\n",
            "    ftabLen: 1048577\n",
            "    ftabSz: 4194308\n",
            "    offsLen: 290104\n",
            "    offsSz: 1160416\n",
            "    lineSz: 64\n",
            "    sideSz: 64\n",
            "    sideBwtSz: 48\n",
            "    sideBwtLen: 192\n",
            "    numSides: 24176\n",
            "    numLines: 24176\n",
            "    ebwtTotLen: 1547264\n",
            "    ebwtTotSz: 1547264\n",
            "    color: 0\n",
            "    reverse: 0\n",
            "Total time for call to driver() for forward index: 00:00:03\n",
            "Reading reference sizes\n",
            "  Time reading reference sizes: 00:00:00\n",
            "Calculating joined length\n",
            "Writing header\n",
            "Reserving space for joined string\n",
            "Joining reference sequences\n",
            "  Time to join reference sequences: 00:00:00\n",
            "  Time to reverse reference sequence: 00:00:00\n",
            "bmax according to bmaxDivN setting: 1160413\n",
            "Using parameters --bmax 870310 --dcv 1024\n",
            "  Doing ahead-of-time memory usage test\n",
            "  Passed!  Constructing with these parameters: --bmax 870310 --dcv 1024\n",
            "Constructing suffix-array element generator\n",
            "Building DifferenceCoverSample\n",
            "  Building sPrime\n",
            "  Building sPrimeOrder\n",
            "  V-Sorting samples\n",
            "  V-Sorting samples time: 00:00:00\n",
            "  Allocating rank array\n",
            "  Ranking v-sort output\n",
            "  Ranking v-sort output time: 00:00:00\n",
            "  Invoking Larsson-Sadakane on ranks\n",
            "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
            "  Sanity-checking and returning\n",
            "Building samples\n",
            "Reserving space for 12 sample suffixes\n",
            "Generating random suffixes\n",
            "QSorting 12 sample offsets, eliminating duplicates\n",
            "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
            "Multikey QSorting 12 samples\n",
            "  (Using difference cover)\n",
            "  Multikey QSorting samples time: 00:00:00\n",
            "Calculating bucket sizes\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 2, merged 7; iterating...\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 1, merged 1; iterating...\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Avg bucket size: 580206 (target: 870309)\n",
            "Converting suffix-array elements to index image\n",
            "Allocating ftab, absorbFtab\n",
            "Entering Ebwt loop\n",
            "Getting block 1 of 8\n",
            "  Reserving size (870310) for bucket 1\n",
            "  Calculating Z arrays for bucket 1\n",
            "  Entering block accumulator loop for bucket 1:\n",
            "  bucket 1: 10%\n",
            "  bucket 1: 20%\n",
            "  bucket 1: 30%\n",
            "  bucket 1: 40%\n",
            "  bucket 1: 50%\n",
            "  bucket 1: 60%\n",
            "  bucket 1: 70%\n",
            "  bucket 1: 80%\n",
            "  bucket 1: 90%\n",
            "  bucket 1: 100%\n",
            "  Sorting block of length 696384 for bucket 1\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:00\n",
            "Returning block of 696385 for bucket 1\n",
            "Getting block 2 of 8\n",
            "  Reserving size (870310) for bucket 2\n",
            "  Calculating Z arrays for bucket 2\n",
            "  Entering block accumulator loop for bucket 2:\n",
            "  bucket 2: 10%\n",
            "  bucket 2: 20%\n",
            "  bucket 2: 30%\n",
            "  bucket 2: 40%\n",
            "  bucket 2: 50%\n",
            "  bucket 2: 60%\n",
            "  bucket 2: 70%\n",
            "  bucket 2: 80%\n",
            "  bucket 2: 90%\n",
            "  bucket 2: 100%\n",
            "  Sorting block of length 333881 for bucket 2\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:00\n",
            "Returning block of 333882 for bucket 2\n",
            "Getting block 3 of 8\n",
            "  Reserving size (870310) for bucket 3\n",
            "  Calculating Z arrays for bucket 3\n",
            "  Entering block accumulator loop for bucket 3:\n",
            "  bucket 3: 10%\n",
            "  bucket 3: 20%\n",
            "  bucket 3: 30%\n",
            "  bucket 3: 40%\n",
            "  bucket 3: 50%\n",
            "  bucket 3: 60%\n",
            "  bucket 3: 70%\n",
            "  bucket 3: 80%\n",
            "  bucket 3: 90%\n",
            "  bucket 3: 100%\n",
            "  Sorting block of length 565798 for bucket 3\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:00\n",
            "Returning block of 565799 for bucket 3\n",
            "Getting block 4 of 8\n",
            "  Reserving size (870310) for bucket 4\n",
            "  Calculating Z arrays for bucket 4\n",
            "  Entering block accumulator loop for bucket 4:\n",
            "  bucket 4: 10%\n",
            "  bucket 4: 20%\n",
            "  bucket 4: 30%\n",
            "  bucket 4: 40%\n",
            "  bucket 4: 50%\n",
            "  bucket 4: 60%\n",
            "  bucket 4: 70%\n",
            "  bucket 4: 80%\n",
            "  bucket 4: 90%\n",
            "  bucket 4: 100%\n",
            "  Sorting block of length 312583 for bucket 4\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:01\n",
            "Returning block of 312584 for bucket 4\n",
            "Getting block 5 of 8\n",
            "  Reserving size (870310) for bucket 5\n",
            "  Calculating Z arrays for bucket 5\n",
            "  Entering block accumulator loop for bucket 5:\n",
            "  bucket 5: 10%\n",
            "  bucket 5: 20%\n",
            "  bucket 5: 30%\n",
            "  bucket 5: 40%\n",
            "  bucket 5: 50%\n",
            "  bucket 5: 60%\n",
            "  bucket 5: 70%\n",
            "  bucket 5: 80%\n",
            "  bucket 5: 90%\n",
            "  bucket 5: 100%\n",
            "  Sorting block of length 783371 for bucket 5\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:00\n",
            "Returning block of 783372 for bucket 5\n",
            "Getting block 6 of 8\n",
            "  Reserving size (870310) for bucket 6\n",
            "  Calculating Z arrays for bucket 6\n",
            "  Entering block accumulator loop for bucket 6:\n",
            "  bucket 6: 10%\n",
            "  bucket 6: 20%\n",
            "  bucket 6: 30%\n",
            "  bucket 6: 40%\n",
            "  bucket 6: 50%\n",
            "  bucket 6: 60%\n",
            "  bucket 6: 70%\n",
            "  bucket 6: 80%\n",
            "  bucket 6: 90%\n",
            "  bucket 6: 100%\n",
            "  Sorting block of length 823242 for bucket 6\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:00\n",
            "Returning block of 823243 for bucket 6\n",
            "Getting block 7 of 8\n",
            "  Reserving size (870310) for bucket 7\n",
            "  Calculating Z arrays for bucket 7\n",
            "  Entering block accumulator loop for bucket 7:\n",
            "  bucket 7: 10%\n",
            "  bucket 7: 20%\n",
            "  bucket 7: 30%\n",
            "  bucket 7: 40%\n",
            "  bucket 7: 50%\n",
            "  bucket 7: 60%\n",
            "  bucket 7: 70%\n",
            "  bucket 7: 80%\n",
            "  bucket 7: 90%\n",
            "  bucket 7: 100%\n",
            "  Sorting block of length 609090 for bucket 7\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:00\n",
            "Returning block of 609091 for bucket 7\n",
            "Getting block 8 of 8\n",
            "  Reserving size (870310) for bucket 8\n",
            "  Calculating Z arrays for bucket 8\n",
            "  Entering block accumulator loop for bucket 8:\n",
            "  bucket 8: 10%\n",
            "  bucket 8: 20%\n",
            "  bucket 8: 30%\n",
            "  bucket 8: 40%\n",
            "  bucket 8: 50%\n",
            "  bucket 8: 60%\n",
            "  bucket 8: 70%\n",
            "  bucket 8: 80%\n",
            "  bucket 8: 90%\n",
            "  bucket 8: 100%\n",
            "  Sorting block of length 517296 for bucket 8\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:00:01\n",
            "Returning block of 517297 for bucket 8\n",
            "Exited Ebwt loop\n",
            "fchr[A]: 0\n",
            "fchr[C]: 1142742\n",
            "fchr[G]: 2322833\n",
            "fchr[T]: 3500270\n",
            "fchr[$]: 4641652\n",
            "Exiting Ebwt::buildToDisk()\n",
            "Returning from initFromVector\n",
            "Wrote 5741800 bytes to primary EBWT file: ecoli_ref.rev.1.bt2\n",
            "Wrote 1160420 bytes to secondary EBWT file: ecoli_ref.rev.2.bt2\n",
            "Re-opening _in1 and _in2 as input streams\n",
            "Returning from Ebwt constructor\n",
            "Headers:\n",
            "    len: 4641652\n",
            "    bwtLen: 4641653\n",
            "    sz: 1160413\n",
            "    bwtSz: 1160414\n",
            "    lineRate: 6\n",
            "    offRate: 4\n",
            "    offMask: 0xfffffff0\n",
            "    ftabChars: 10\n",
            "    eftabLen: 20\n",
            "    eftabSz: 80\n",
            "    ftabLen: 1048577\n",
            "    ftabSz: 4194308\n",
            "    offsLen: 290104\n",
            "    offsSz: 1160416\n",
            "    lineSz: 64\n",
            "    sideSz: 64\n",
            "    sideBwtSz: 48\n",
            "    sideBwtLen: 192\n",
            "    numSides: 24176\n",
            "    numLines: 24176\n",
            "    ebwtTotLen: 1547264\n",
            "    ebwtTotSz: 1547264\n",
            "    color: 0\n",
            "    reverse: 1\n",
            "Total time for backward call to driver() for mirror index: 00:00:03\n"
          ]
        }
      ],
      "source": [
        "# Faça o download do genoma de referencia que usaremos: E. coli K-12 genoma de referencia do NCBI (https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/).\n",
        "# Download E. coli genoma de referencia (K-12 cepa)\n",
        "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz\n",
        "!gunzip GCF_000005845.2_ASM584v2_genomic.fna.gz\n",
        "\n",
        "# Renomeando\n",
        "!mv GCF_000005845.2_ASM584v2_genomic.fna ecoli_ref.fna\n",
        "\n",
        "# Indexação do genoma de referencia com o Bowtie2\n",
        "!bowtie2-build ecoli_ref.fna ecoli_ref"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "GFW6AxY-s8LJ",
        "outputId": "cad699a7-7bee-4ccf-c3e1-ee6777ee2a09"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  blends-common libkdf5-2 libncbi-vdb2 libncbi-wvdb2 med-config menu\n",
            "Suggested packages:\n",
            "  blends-doc menu-l10n gksu | kde-runtime | ktsuss\n",
            "The following NEW packages will be installed:\n",
            "  blends-common libkdf5-2 libncbi-vdb2 libncbi-wvdb2 med-config menu sra-toolkit\n",
            "0 upgraded, 7 newly installed, 0 to remove and 32 not upgraded.\n",
            "Need to get 8,290 kB of archives.\n",
            "After this operation, 23.0 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 menu amd64 2.1.47ubuntu4 [354 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 blends-common all 0.7.4ubuntu1 [15.9 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libkdf5-2 amd64 2.11.2+dfsg-4build2 [14.7 kB]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libncbi-vdb2 amd64 2.11.2+dfsg-4build2 [1,364 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libncbi-wvdb2 amd64 2.11.2+dfsg-4build2 [1,252 kB]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy/universe amd64 sra-toolkit amd64 2.11.3+dfsg-1ubuntu1 [5,277 kB]\n",
            "Get:7 http://archive.ubuntu.com/ubuntu jammy/universe amd64 med-config all 3.7.1 [11.6 kB]\n",
            "Fetched 8,290 kB in 1s (9,375 kB/s)\n",
            "Preconfiguring packages ...\n",
            "Selecting previously unselected package menu.\n",
            "(Reading database ... 125169 files and directories currently installed.)\n",
            "Preparing to unpack .../0-menu_2.1.47ubuntu4_amd64.deb ...\n",
            "Unpacking menu (2.1.47ubuntu4) ...\n",
            "Selecting previously unselected package blends-common.\n",
            "Preparing to unpack .../1-blends-common_0.7.4ubuntu1_all.deb ...\n",
            "Unpacking blends-common (0.7.4ubuntu1) ...\n",
            "Selecting previously unselected package libkdf5-2.\n",
            "Preparing to unpack .../2-libkdf5-2_2.11.2+dfsg-4build2_amd64.deb ...\n",
            "Unpacking libkdf5-2 (2.11.2+dfsg-4build2) ...\n",
            "Selecting previously unselected package libncbi-vdb2.\n",
            "Preparing to unpack .../3-libncbi-vdb2_2.11.2+dfsg-4build2_amd64.deb ...\n",
            "Unpacking libncbi-vdb2 (2.11.2+dfsg-4build2) ...\n",
            "Selecting previously unselected package libncbi-wvdb2.\n",
            "Preparing to unpack .../4-libncbi-wvdb2_2.11.2+dfsg-4build2_amd64.deb ...\n",
            "Unpacking libncbi-wvdb2 (2.11.2+dfsg-4build2) ...\n",
            "Selecting previously unselected package sra-toolkit.\n",
            "Preparing to unpack .../5-sra-toolkit_2.11.3+dfsg-1ubuntu1_amd64.deb ...\n",
            "Unpacking sra-toolkit (2.11.3+dfsg-1ubuntu1) ...\n",
            "Selecting previously unselected package med-config.\n",
            "Preparing to unpack .../6-med-config_3.7.1_all.deb ...\n",
            "Unpacking med-config (3.7.1) ...\n",
            "Setting up libncbi-vdb2 (2.11.2+dfsg-4build2) ...\n",
            "Setting up libkdf5-2 (2.11.2+dfsg-4build2) ...\n",
            "Setting up libncbi-wvdb2 (2.11.2+dfsg-4build2) ...\n",
            "Setting up menu (2.1.47ubuntu4) ...\n",
            "Setting up blends-common (0.7.4ubuntu1) ...\n",
            "Setting up sra-toolkit (2.11.3+dfsg-1ubuntu1) ...\n",
            "Setting up med-config (3.7.1) ...\n",
            "Adding group `med' (GID 107) ...\n",
            "Done.\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.8) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "Processing triggers for menu (2.1.47ubuntu4) ...\n",
            "\n",
            "2025-03-17T20:07:29 prefetch.2.11.3: Current preference is set to retrieve SRA Normalized Format files with full base quality scores.\n",
            "2025-03-17T20:07:29 prefetch.2.11.3: 1) Downloading 'SRR12345678'...\n",
            "2025-03-17T20:07:29 prefetch.2.11.3: SRA Normalized Format file is being retrieved, if this is different from your preference, it may be due to current file availability.\n",
            "2025-03-17T20:07:29 prefetch.2.11.3:  Downloading via HTTPS...\n",
            "2025-03-17T20:07:30 prefetch.2.11.3:  HTTPS download succeed\n",
            "2025-03-17T20:07:30 prefetch.2.11.3:  'SRR12345678' is valid\n",
            "2025-03-17T20:07:30 prefetch.2.11.3: 1) 'SRR12345678' was downloaded successfully\n",
            "2025-03-17T20:07:30 prefetch.2.11.3: 'SRR12345678' has 0 unresolved dependencies\n",
            "spots read      : 23,655\n",
            "reads read      : 47,310\n",
            "reads written   : 47,310\n"
          ]
        }
      ],
      "source": [
        "# E. coli disponível publicamente do Sequence Read Archive (SRA); SRA SRR12345678\n",
        "# Instale o SRA toolkit para fazer o download da amostra\n",
        "!apt install -y sra-toolkit\n",
        "\n",
        "# Download da amostra\n",
        "!prefetch SRR12345678\n",
        "\n",
        "# A amostra será baixada como arquivos FASTQ (por exemplo, SRR12345678_1.fastq e SRR12345678_2.fastq para leituras de extremidade pareada)\n",
        "!fasterq-dump SRR12345678"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "M_OvuevXuEzd",
        "outputId": "af3d399e-a93b-4b8d-cf18-0c8e56745532"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "23655 reads; of these:\n",
            "  23655 (100.00%) were paired; of these:\n",
            "    23526 (99.45%) aligned concordantly 0 times\n",
            "    0 (0.00%) aligned concordantly exactly 1 time\n",
            "    129 (0.55%) aligned concordantly >1 times\n",
            "    ----\n",
            "    23526 pairs aligned concordantly 0 times; of these:\n",
            "      0 (0.00%) aligned discordantly 1 time\n",
            "    ----\n",
            "    23526 pairs aligned 0 times concordantly or discordantly; of these:\n",
            "      47052 mates make up the pairs; of these:\n",
            "        46461 (98.74%) aligned 0 times\n",
            "        0 (0.00%) aligned exactly 1 time\n",
            "        591 (1.26%) aligned >1 times\n",
            "1.79% overall alignment rate\n"
          ]
        }
      ],
      "source": [
        "# Run Bowtie2 alinhamento\n",
        "!bowtie2 -x ecoli_ref -1 SRR12345678_1.fastq -2 SRR12345678_2.fastq -S aligned_output.sam\n",
        "\n",
        "# Converção do formato SAM para BAM\n",
        "!samtools view -bS aligned_output.sam -o aligned_output.bam\n",
        "\n",
        "# 'Sort' do arquivo BAM\n",
        "!samtools sort aligned_output.bam -o aligned_output_sorted.bam\n",
        "\n",
        "# Indexação do arquivo BAM organizado 'sorted'\n",
        "!samtools index aligned_output_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "a-c4Bv9xustp",
        "outputId": "8ff14b5f-5454-4590-f85f-110cd60c7567"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "SRR12345678.200\t73\tNC_000913.3\t224304\t0\t151M\tNC_000913.3\t224304\t0\tTACGTAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGTGCGCAGGCGGTTTTGTAAGACAGATGTGAAATGCCCGGGCTTAACCTGGGAACTGCATTTGTGACTGCAAGGCTAGAATCTGGCAGAGGGGGGTAGAATTCCACGT\tarray('B', [34, 34, 34, 34, 34, 37, 37, 34, 34, 37, 33, 32, 38, 38, 38, 38, 38, 38, 38, 38, 37, 38, 39, 38, 38, 38, 36, 37, 39, 39, 39, 39, 38, 39, 39, 39, 38, 38, 38, 38, 39, 39, 39, 38, 38, 35, 36, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 16, 33, 38, 39, 39, 39, 39, 18, 37, 39, 39, 39, 38, 37, 38, 39, 39, 39, 39, 36, 39, 39, 36, 31, 36, 38, 38, 39, 39, 39, 39, 39, 39, 30, 38, 37, 38, 39, 38, 39, 39, 39, 37, 38, 39, 39, 38, 38, 17, 38, 39, 37, 38, 39, 37, 38, 17, 37, 36, 38, 37, 39, 39, 39, 37, 37, 39, 37, 39, 39, 39, 38, 38, 36, 37, 32, 34, 30, 37, 26, 32, 36, 37, 36, 37, 37, 37, 37, 33, 37, 35, 36])\t[('AS', -89), ('XN', 0), ('XM', 19), ('XO', 0), ('XG', 0), ('NM', 19), ('MD', '4G39C0A12G0T4T12C8C16C2A0T4G0C1A3T2G3C1T19G2'), ('YT', 'UP')]\n",
            "SRR12345678.200\t133\tNC_000913.3\t224304\t0\tNone\tNC_000913.3\t224304\t0\tCCTGTTTGCTCCCCACGCTTTCGTGCATGAGCGTCAATCTTAGCCCAGGGGGCTGCCTTCGCCATCGGTATTCCTCCACATATCTACGCATTTCACTGCTACACGTGGAATTCTACCCCCCTCTGCCAGATTCTAGCCTTGCAGTCCCCCC\tarray('B', [33, 32, 33, 34, 33, 37, 37, 37, 37, 37, 37, 37, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 19, 38, 20, 37, 38, 37, 38, 38, 38, 38, 32, 37, 37, 38, 39, 33, 35, 38, 38, 18, 34, 38, 38, 16, 36, 32, 36, 38, 38, 38, 39, 38, 38, 33, 36, 36, 38, 38, 37, 39, 37, 38, 37, 39, 37, 39, 39, 39, 38, 36, 38, 17, 38, 38, 39, 39, 39, 39, 39, 39, 39, 38, 38, 38, 35, 38, 37, 39, 39, 39, 35, 38, 38, 37, 38, 39, 39, 39, 35, 39, 39, 38, 37, 39, 39, 39, 39, 39, 35, 39, 39, 39, 39, 38, 38, 38, 38, 39, 37, 38, 37, 34, 16, 30, 38, 37, 37, 39, 39, 37, 33, 38, 37, 39, 39, 39, 39, 37, 38, 39, 27, 27, 13, 27, 12])\t[('YT', 'UP')]\n",
            "SRR12345678.429\t73\tNC_000913.3\t224304\t0\t151M\tNC_000913.3\t224304\t0\tTACGTAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGTGCGCAGGCGGTCCGCTAAGACAGATGTGAAATCCCCGGGCTTAACCTGGGAACTGCATTTGTGACTGGCGGGCTAGAGTATGGCAGAGGGGGGTAGAATTCCACGT\tarray('B', [32, 32, 32, 18, 32, 37, 37, 33, 33, 37, 33, 33, 38, 38, 38, 37, 36, 38, 38, 36, 17, 37, 38, 38, 38, 30, 34, 37, 39, 39, 37, 39, 39, 39, 39, 38, 38, 38, 38, 38, 39, 39, 38, 38, 38, 38, 38, 38, 34, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 34, 38, 39, 37, 39, 37, 39, 39, 39, 39, 39, 39, 39, 38, 39, 39, 37, 39, 39, 38, 38, 38, 38, 38, 38, 39, 39, 37, 38, 39, 39, 39, 39, 38, 39, 39, 38, 39, 38, 39, 39, 39, 39, 39, 39, 38, 39, 39, 39, 39, 39, 39, 38, 38, 38, 38, 38, 38, 39, 39, 39, 39, 39, 36, 39, 38, 39, 32, 39, 35, 37, 39, 38, 38, 38, 38, 33, 24, 35, 30, 37, 33, 37, 37, 33, 37, 37, 37, 37, 32])\t[('AS', -90), ('XN', 0), ('XM', 18), ('XO', 0), ('XG', 0), ('NM', 18), ('MD', '4G39C0A10T0T1T4T21C16C2A0T6A0A3T4C1C1T19G2'), ('YT', 'UP')]\n",
            "SRR12345678.429\t133\tNC_000913.3\t224304\t0\tNone\tNC_000913.3\t224304\t0\tCCTGTTTGCTCCCCACGCTTTCGTGCATGAGCGTCAGTATTGGCCCAGGGGGCTGCCTTCGCCATCGGTATTCCTCCACATCTCTACGCATTTCACTGCTACACGTGGAATTCTACCCCCCTCTGCCATACTCTAGCCCGCCAGTCCCTTT\tarray('B', [31, 32, 30, 32, 32, 37, 37, 37, 33, 37, 35, 37, 34, 38, 33, 36, 34, 37, 38, 37, 38, 38, 38, 37, 32, 37, 35, 38, 35, 38, 39, 37, 38, 38, 36, 37, 38, 39, 38, 39, 39, 38, 18, 36, 37, 17, 17, 32, 34, 16, 15, 32, 36, 36, 37, 38, 37, 37, 38, 39, 16, 36, 37, 36, 32, 38, 16, 30, 36, 38, 37, 39, 39, 37, 38, 31, 37, 37, 39, 37, 38, 38, 39, 39, 39, 38, 39, 38, 38, 36, 30, 37, 39, 37, 39, 39, 37, 37, 37, 37, 33, 38, 38, 17, 36, 30, 37, 37, 30, 37, 38, 39, 39, 39, 37, 39, 39, 36, 37, 34, 35, 34, 38, 39, 33, 38, 38, 37, 16, 34, 38, 38, 39, 33, 35, 35, 16, 27, 37, 31, 30, 35, 38, 30, 38, 38, 16, 16, 16, 15, 15])\t[('YT', 'UP')]\n",
            "SRR12345678.675\t73\tNC_000913.3\t224304\t0\t151M\tNC_000913.3\t224304\t0\tTACGTAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGTGCGCAGGCGGTTCGTTAAGACAGATGTGAAATCCCCGGGCTTAACCTGGGAACTGCATTTGTGACTGGCGAGCTAGAGTATGGCAGAGGGGGGTAGAATTCCACGT\tarray('B', [34, 34, 34, 34, 33, 37, 37, 34, 34, 37, 34, 34, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 39, 38, 38, 38, 38, 38, 39, 39, 38, 39, 39, 39, 39, 39, 38, 38, 38, 38, 39, 39, 39, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 37, 38, 36, 37, 38, 34, 38, 33, 33, 38, 37, 39, 38, 39, 39, 38, 39, 39, 39, 39, 37, 39, 39, 38, 38, 38, 38, 38, 38, 39, 39, 39, 39, 39, 39, 39, 39, 38, 39, 39, 38, 39, 39, 39, 39, 39, 39, 39, 38, 39, 39, 38, 39, 39, 39, 39, 39, 38, 35, 38, 38, 38, 39, 39, 39, 39, 37, 39, 38, 37, 39, 39, 38, 38, 37, 39, 38, 38, 38, 38, 37, 33, 35, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 36])\t[('AS', -80), ('XN', 0), ('XM', 16), ('XO', 0), ('XG', 0), ('NM', 16), ('MD', '4G7A31C0A11T6T21C16C2A0T6A4T4C1C1T19G2'), ('YT', 'UP')]\n"
          ]
        }
      ],
      "source": [
        "# Visualização do alinhamento\n",
        "import pysam\n",
        "\n",
        "# Open the sorted BAM file\n",
        "bamfile = pysam.AlignmentFile(\"aligned_output_sorted.bam\", \"rb\")\n",
        "\n",
        "# Print the first few alignments\n",
        "for read in bamfile.head(5):\n",
        "    print(read)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ck4bTLfMvEyT",
        "outputId": "701afc46-1f94-4909-9782-6f551d7409e8"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "47310 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "47310 + 0 primary\n",
            "0 + 0 secondary\n",
            "0 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "0 + 0 primary duplicates\n",
            "849 + 0 mapped (1.79% : N/A)\n",
            "849 + 0 primary mapped (1.79% : N/A)\n",
            "47310 + 0 paired in sequencing\n",
            "23655 + 0 read1\n",
            "23655 + 0 read2\n",
            "258 + 0 properly paired (0.55% : N/A)\n",
            "258 + 0 with itself and mate mapped\n",
            "591 + 0 singletons (1.25% : N/A)\n",
            "0 + 0 with mate mapped to a different chr\n",
            "0 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ]
        }
      ],
      "source": [
        "# Calculando a estatística do alinhamento\n",
        "\n",
        "!samtools flagstat aligned_output_sorted.bam"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Imqu1AkOv0De",
        "outputId": "65a566d5-77f3-431e-dcaa-7b3740ddc2a6"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "[mpileup] 1 samples in 1 input files\n"
          ]
        }
      ],
      "source": [
        "# Gerando o arquivo pileup\n",
        "!samtools mpileup -f ecoli_ref.fna aligned_output_sorted.bam > alignment.pileup"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "kDyy_6xFwA9P",
        "outputId": "999e11a2-01c1-4d9c-ed82-4ca82d762c1b"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Note: none of --samples-file, --ploidy or --ploidy-file given, assuming all sites are diploid\n",
            "[mpileup] 1 samples in 1 input files\n",
            "[mpileup] maximum number of reads per input file set to -d 250\n"
          ]
        }
      ],
      "source": [
        "# Chamada de variantes: gerando o arquivo VCF\n",
        "!bcftools mpileup -f ecoli_ref.fna aligned_output_sorted.bam | bcftools call -mv -o variants.vcf"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "o-zJAfqAwNvY",
        "outputId": "d970e4f2-6fa5-4c2f-d770-953bc87635c2"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "##fileformat=VCFv4.2\n",
            "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
            "##bcftoolsVersion=1.13+htslib-1.13+ds\n",
            "##bcftoolsCommand=mpileup -f ecoli_ref.fna aligned_output_sorted.bam\n",
            "##reference=file://ecoli_ref.fna\n",
            "##contig=<ID=NC_000913.3,length=4641652>\n",
            "##ALT=<ID=*,Description=\"Represents allele(s) other than observed.\">\n",
            "##INFO=<ID=INDEL,Number=0,Type=Flag,Description=\"Indicates that the variant is an INDEL.\">\n",
            "##INFO=<ID=IDV,Number=1,Type=Integer,Description=\"Maximum number of raw reads supporting an indel\">\n",
            "##INFO=<ID=IMF,Number=1,Type=Float,Description=\"Maximum fraction of raw reads supporting an indel\">\n",
            "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Raw read depth\">\n",
            "##INFO=<ID=VDB,Number=1,Type=Float,Description=\"Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)\",Version=\"3\">\n",
            "##INFO=<ID=RPBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Read Position Bias (closer to 0 is better)\">\n",
            "##INFO=<ID=MQBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Mapping Quality Bias (closer to 0 is better)\">\n",
            "##INFO=<ID=BQBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Base Quality Bias (closer to 0 is better)\">\n",
            "##INFO=<ID=MQSBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Mapping Quality vs Strand Bias (closer to 0 is better)\">\n",
            "##INFO=<ID=SCBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Soft-Clip Length Bias (closer to 0 is better)\">\n",
            "##INFO=<ID=FS,Number=1,Type=Float,Description=\"Phred-scaled p-value using Fisher's exact test to detect strand bias\">\n",
            "##INFO=<ID=SGB,Number=1,Type=Float,Description=\"Segregation based metric.\">\n",
            "##INFO=<ID=MQ0F,Number=1,Type=Float,Description=\"Fraction of MQ0 reads (smaller is better)\">\n",
            "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"List of Phred-scaled genotype likelihoods\">\n",
            "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
            "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
            "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">\n",
            "##INFO=<ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
            "##INFO=<ID=MQ,Number=1,Type=Integer,Description=\"Average mapping quality\">\n",
            "##bcftools_callVersion=1.13+htslib-1.13+ds\n",
            "##bcftools_callCommand=call -mv -o variants.vcf; Date=Mon Mar 17 20:16:27 2025\n",
            "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\taligned_output_sorted.bam\n",
            "NC_000913.3\t2730476\t.\tC\tG\t5.75677\t.\tDP=104;VDB=1;SGB=-0.693147;FS=0;MQ0F=0.5;AC=2;AN=2;DP4=0,0,23,29;MQ=0\tGT:PL\t1/1:34,157,0\n"
          ]
        }
      ],
      "source": [
        "!head -n 30 variants.vcf"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "CFpuKf5Twmoi"
      },
      "outputs": [],
      "source": [
        "# Filtrando o arquivo VCF\n",
        "!bcftools filter -i 'QUAL > 3' variants.vcf -o filtered_variants.vcf"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "GwSyP_Z0wvC3",
        "outputId": "638f5814-f7e4-4b82-ce6a-4f7eb26066c2"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "##fileformat=VCFv4.2\n",
            "##FILTER=<ID=PASS,Description=\"All filters passed\">\n",
            "##bcftoolsVersion=1.13+htslib-1.13+ds\n",
            "##bcftoolsCommand=mpileup -f ecoli_ref.fna aligned_output_sorted.bam\n",
            "##reference=file://ecoli_ref.fna\n",
            "##contig=<ID=NC_000913.3,length=4641652>\n",
            "##ALT=<ID=*,Description=\"Represents allele(s) other than observed.\">\n",
            "##INFO=<ID=INDEL,Number=0,Type=Flag,Description=\"Indicates that the variant is an INDEL.\">\n",
            "##INFO=<ID=IDV,Number=1,Type=Integer,Description=\"Maximum number of raw reads supporting an indel\">\n",
            "##INFO=<ID=IMF,Number=1,Type=Float,Description=\"Maximum fraction of raw reads supporting an indel\">\n",
            "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Raw read depth\">\n",
            "##INFO=<ID=VDB,Number=1,Type=Float,Description=\"Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)\",Version=\"3\">\n",
            "##INFO=<ID=RPBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Read Position Bias (closer to 0 is better)\">\n",
            "##INFO=<ID=MQBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Mapping Quality Bias (closer to 0 is better)\">\n",
            "##INFO=<ID=BQBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Base Quality Bias (closer to 0 is better)\">\n",
            "##INFO=<ID=MQSBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Mapping Quality vs Strand Bias (closer to 0 is better)\">\n",
            "##INFO=<ID=SCBZ,Number=1,Type=Float,Description=\"Mann-Whitney U-z test of Soft-Clip Length Bias (closer to 0 is better)\">\n",
            "##INFO=<ID=FS,Number=1,Type=Float,Description=\"Phred-scaled p-value using Fisher's exact test to detect strand bias\">\n",
            "##INFO=<ID=SGB,Number=1,Type=Float,Description=\"Segregation based metric.\">\n",
            "##INFO=<ID=MQ0F,Number=1,Type=Float,Description=\"Fraction of MQ0 reads (smaller is better)\">\n",
            "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"List of Phred-scaled genotype likelihoods\">\n",
            "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
            "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes for each ALT allele, in the same order as listed\">\n",
            "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">\n",
            "##INFO=<ID=DP4,Number=4,Type=Integer,Description=\"Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases\">\n",
            "##INFO=<ID=MQ,Number=1,Type=Integer,Description=\"Average mapping quality\">\n",
            "##bcftools_callVersion=1.13+htslib-1.13+ds\n",
            "##bcftools_callCommand=call -mv -o variants.vcf; Date=Mon Mar 17 20:16:27 2025\n",
            "##bcftools_filterVersion=1.13+htslib-1.13+ds\n",
            "##bcftools_filterCommand=filter -i 'QUAL > 3' -o filtered_variants.vcf variants.vcf; Date=Mon Mar 17 20:20:35 2025\n",
            "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\taligned_output_sorted.bam\n",
            "NC_000913.3\t2730476\t.\tC\tG\t5.75677\tPASS\tDP=104;VDB=1;SGB=-0.693147;FS=0;MQ0F=0.5;AC=2;AN=2;DP4=0,0,23,29;MQ=0\tGT:PL\t1/1:34,157,0\n",
            "NC_000913.3\t2730496\t.\tA\tG\t3.78127\tPASS\tDP=104;VDB=1;SGB=-0.693147;FS=0;MQ0F=0.5;AC=2;AN=2;DP4=0,0,24,28;MQ=0\tGT:PL\t1/1:31,154,0\n",
            "NC_000913.3\t2730498\t.\tG\tC\t5.75677\tPASS\tDP=104;VDB=1;SGB=-0.693147;FS=0;MQ0F=0.5;AC=2;AN=2;DP4=0,0,23,29;MQ=0\tGT:PL\t1/1:34,157,0\n",
            "NC_000913.3\t2730500\t.\tG\tT\t3.78393\tPASS\tDP=104;VDB=1;SGB=-0.693147;FS=0;MQ0F=0.5;AC=2;AN=2;DP4=0,0,23,29;MQ=0\tGT:PL\t1/1:31,151,0\n",
            "NC_000913.3\t2730505\t.\tA\tT\t5.75677\tPASS\tDP=104;VDB=1;SGB=-0.693147;FS=0;MQ0F=0.5;AC=2;AN=2;DP4=0,0,23,29;MQ=0\tGT:PL\t1/1:34,157,0\n",
            "NC_000913.3\t2730510\t.\tT\tC\t5.70756\tPASS\tDP=104;VDB=1;SGB=-0.693147;RPBZ=-1.12288;BQBZ=2.17321;FS=0;MQ0F=0.5;AC=2;AN=2;DP4=1,0,22,29;MQ=0\tGT:PL\t1/1:31,136,0\n",
            "NC_000913.3\t2730517\t.\tA\tC\t12.5751\tPASS\tDP=104;VDB=0.999824;SGB=-0.693146;RPBZ=-1.28338;BQBZ=0.856753;FS=0;MQ0F=0.432692;AC=2;AN=2;DP4=1,0,16,28;MQ=0\tGT:PL\t1/1:39,119,0\n"
          ]
        }
      ],
      "source": [
        "!head -n 38 filtered_variants.vcf"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xWKnKjhXxIK8",
        "outputId": "1843007d-216c-4560-e035-88b34e020847"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "NC_000913.3\t2730476\t.\tC\tG\t5.75677\tPASS\tDP=104;VDB=1;SGB=-0.693147;FS=0;MQ0F=0.5;AC=2;AN=2;DP4=0,0,23,29;MQ=0\tGT:PL\t1/1:34,157,0\n",
            "\n",
            "NC_000913.3\t2730496\t.\tA\tG\t3.78127\tPASS\tDP=104;VDB=1;SGB=-0.693147;FS=0;MQ0F=0.5;AC=2;AN=2;DP4=0,0,24,28;MQ=0\tGT:PL\t1/1:31,154,0\n",
            "\n",
            "NC_000913.3\t2730498\t.\tG\tC\t5.75677\tPASS\tDP=104;VDB=1;SGB=-0.693147;FS=0;MQ0F=0.5;AC=2;AN=2;DP4=0,0,23,29;MQ=0\tGT:PL\t1/1:34,157,0\n",
            "\n",
            "NC_000913.3\t2730500\t.\tG\tT\t3.78393\tPASS\tDP=104;VDB=1;SGB=-0.693147;FS=0;MQ0F=0.5;AC=2;AN=2;DP4=0,0,23,29;MQ=0\tGT:PL\t1/1:31,151,0\n",
            "\n",
            "NC_000913.3\t2730505\t.\tA\tT\t5.75677\tPASS\tDP=104;VDB=1;SGB=-0.693147;FS=0;MQ0F=0.5;AC=2;AN=2;DP4=0,0,23,29;MQ=0\tGT:PL\t1/1:34,157,0\n",
            "\n",
            "NC_000913.3\t2730510\t.\tT\tC\t5.70756\tPASS\tDP=104;VDB=1;SGB=-0.693147;RPBZ=-1.12288;BQBZ=2.17321;FS=0;MQ0F=0.5;AC=2;AN=2;DP4=1,0,22,29;MQ=0\tGT:PL\t1/1:31,136,0\n",
            "\n",
            "NC_000913.3\t2730517\t.\tA\tC\t12.5751\tPASS\tDP=104;VDB=0.999824;SGB=-0.693146;RPBZ=-1.28338;BQBZ=0.856753;FS=0;MQ0F=0.432692;AC=2;AN=2;DP4=1,0,16,28;MQ=0\tGT:PL\t1/1:39,119,0\n",
            "\n",
            "NC_000913.3\t3428122\t.\tA\tC\t9.7001\tPASS\tDP=153;VDB=0.999999;SGB=-0.693147;RPBZ=-1.44263;MQBZ=-6.21733;MQSBZ=-0.342263;BQBZ=1.29526;FS=0;MQ0F=0.379085;AC=2;AN=2;DP4=2,1,21,36;MQ=0\tGT:PL\t1/1:36,135,0\n",
            "\n"
          ]
        }
      ],
      "source": [
        "import pysam\n",
        "\n",
        "# Abra o arquivo VCF:\n",
        "vcf = pysam.VariantFile(\"filtered_variants.vcf\")\n",
        "\n",
        "# Imprima as primeiras variantes\n",
        "for record in vcf:\n",
        "    print(record)\n",
        "    #break  # Remova esta linha para imprimir mais variantes"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4rZZXSapxiNt"
      },
      "source": [
        "# Etapa: Visualizar variantes\n",
        "\n",
        "Você também pode visualizar as variantes no IGV (Integrative Genomics Viewer):\n",
        "\n",
        "Baixe o arquivo variants.vcf do Colab para sua máquina local.\n",
        "\n",
        "Abra o IGV e carregue o genoma de referência (ecoli_ref.fna) e o arquivo VCF."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Referências adicionais\n",
        "https://www.deepseek.com/"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
